cpp/include/raft/core/resource/device_memory_resource.hpp

```diff
@@ -144,7 +177,7 @@ class workspace_resource_factory : public resource_factory {
     // Note, the workspace does not claim all this memory from the start, so it's still usable by
     // the main resource as well.
     // This limit is merely an order for algorithm internals to plan the batching accordingly.
-    return total_size / 2;
+    return total_size / 4;
```

Comment:
The OOM errors we have seen with CAGRA were related to workspace pool grabbing all this place. What about limiting to a much smaller workspace size? (E.g. [faiss has](https://github.com/facebookresearch/faiss/blob/abff75ef078aa7b2258c5ea965f2353691c71b42/faiss/gpu/StandardGpuResources.h#L56) 1.5 GiB limit).

Reply:
That is an option, but so far I think it's not necessary. I also think it can hurt performance a little by reducing the batch size in places like ivf_pq::search or ivf_pq::extend.

With the current proposal, ann-bench executable (as a user of raft) set these resources:
 - default - pool on top of device memory
 - limited workspace - shares the same pool with default
 - large workspace - managed memory (without pooling)

Hence the dataset/user allocations do not conflict for the same memory with the workspace (as they both use the same pool). At the same time, large temporary allocations (such as the cagra graph on device) use the managed memory and free it as soon as the algorithm finishes.